Feature: Implement Docker-based building and client tests

AnXh3L0 opened the pull request:
This PR aims to improve the current Docker-building methodology, by re-using the same build.yml file while adding a new install-docker script. Among the changes, we have implemented parallel Docker building and test execution on all the distros we currently support, with bookworm as the default one.

Additionally, the Cypress tests have been changed to run on these Docker instances, and Lighthouse testing has been added for the three main public-facing pages, that do not require log in.

Currently, the workflows may fail because there's a few Cypress tests failing, but the workflow is generally stable. A few changes have been introduced to the .dockerignore and the .gitignore files, in order to prevent local tests from being accidentally commited to the remote repo.

Moreover, almost every step of the build is commented, in order to make it easier to understand what each line does. At the moment, the workflow takes about 30-40 minutes to execute, when running the default bookworm build process.

AnXh3L0 committed:
feat: implement Docker-based testing for GitHub Actions

AnXh3L0 committed:
fix: run build process inside Docker container - fixes debuild requirement error

AnXh3L0 committed:
fix: replace build script with direct build process

AnXh3L0 committed:
fix: break down Docker build into separate steps

AnXh3L0 committed:
fix: avoid recursive copy issue

AnXh3L0 committed:
fix: add missing Python build dependencies

AnXh3L0 committed:
fix: changed workflow and retest

AnXh3L0 committed:
Removed duplicated items, re-use Docker files

AnXh3L0 committed:
fix: missing branch

AnXh3L0 committed:
fix: fixed wrong path

AnXh3L0 committed:
check for subdirectories as well

AnXh3L0 committed:
Added a new test script to make it work without root inside Docker

AnXh3L0 committed:
Changed dockerignore

AnXh3L0 committed:
Changed order of commands

AnXh3L0 committed:
Re-pull the arguments after they're lost

AnXh3L0 committed:
Added debugging for cypress installation

AnXh3L0 committed:
Added lighthouse auditing

AnXh3L0 committed:
Fix: fixed escaping char

AnXh3L0 committed:
Fix: changed syntax because of GHActions failure

AnXh3L0 committed:
Fix: changed to jq

AnXh3L0 committed:
Fix: Added more pages

AnXh3L0 committed:
Fix: Added missing element

AnXh3L0 committed:
Fix: Make Chrome headless

AnXh3L0 committed:
Fix: fixed missing report string

evilaliv3 commented:
Hey @AnXh3L0 

I just noticed this now, shall i proceed with a first round of review or is there still something you want to complete?

I will give this a try tomorrow.

I see that the test run is failing but it is probably due to problems related to the devel branch, while for this changes we could directly instegrate inside the stable.

AnXh3L0 commented:
@evilaliv3 yes, I think you can start with a first round of review, I don't think there's anything else I can add for now.

evilaliv3 commented:
I remodeled the pull request to be based on current stable branch in order to perform a first review.

evilaliv3 commented on .gitignore:
```diff
@@ -3,3 +3,20 @@ backend/workingdir/
+build/
+
+# Lighthouse reports generated by CI/CD
+client/lighthouse-results/
```
As discussed let's remove thesechanges will we figure out in a separated ticket how to handler .gitignore

evilaliv3 commented on docker/.dockerignore:
```diff
@@ -0,0 +1,14 @@
+*.tmp
+*.log
+.DS_Store
+Thumbs.db 
```
For the moment lets skipp to add a dockerignore file trying to understand better how to handle it. Anyway for example on github there won't be files like .DS_Store or Thumds.db

evilaliv3 commented on docker/Dockerfile:
```diff
@@ -1,16 +1,59 @@
+        /tmp/install-docker.sh; \
+    else \
+        echo "--- Performing remote installation ---"; \
+        wget https://deb.globaleaks.org/install-globaleaks.sh -O /tmp/install-globaleaks.sh && \
```
it was discussed to side load the file locally without downloading it from deb.globaleaks.org

evilaliv3 commented on docker/docker-compose.yml:
```diff
@@ -1,16 +1,147 @@
+    profiles:
+      - testing
+      - test-focal
+      - test-all
```
this seems very doplicate once for each distro/version but if really necessry fine to keep this for now

AnXh3L0 replied:
These are necessary if we want to be able to run multiple/parallel Docker tests for the different distros/versions

evilaliv3 commented on docker/install-docker.sh:
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+# Docker-specific GlobaLeaks installation script
```
Which are the specific details of this script? It appears mostly duplicated.

I suggest to evaluate to amend the original script with the few changes that are necessary to make it work in both the contexts.

evilaliv3 reviewed (changes requested):
Hello @AnXh3L0 , i've loaded all my current comments suggesting the changes to apply before merge.

AnXh3L0 commented:
Thanks for the comments @evilaliv3, I am checking and applying the changes you requested.

AnXh3L0 committed:
Reverted to normal gitignore

AnXh3L0 committed:
Removed dockerignore for now

AnXh3L0 committed:
Modified script to use local install script

AnXh3L0 committed:
Fixed wrong path

AnXh3L0 committed:
Updated path

AnXh3L0 committed:
Use local context

AnXh3L0 committed:
Jump to parent directory

AnXh3L0 committed:
Create empty directory for script to succeed

AnXh3L0 committed:
Removed docker install script

AnXh3L0 committed:
Modified install script to run Docker context

AnXh3L0 committed:
Changed paths

AnXh3L0 committed:
Fixed masking issue

AnXh3L0 committed:
Added debugging to check for Docker env

evilaliv3 commented on docker/docker-compose.yml:
```diff
@@ -1,16 +1,147 @@
+      - 8443:8443
+    profiles:
+      - production
+      - ""  # Default profile
```
@AnXh3L0: which is the difference in these two profiles and when are they used?

AnXh3L0 replied:
It's just for clarity/declaration purposes, it doesn't really do much else. Basically it allows us to test for every distro/case with a single command, and if necessary, it can be re-used in the future to run different things when the specific profile is chosen.

evilaliv3 commented on docker/Dockerfile:
```diff
@@ -1,14 +1,58 @@
+#   DISTRIBUTION:   The target distribution codename (e.g., bookworm)
+#   INSTALL_FROM:   Installation mode. 'remote' (default) or 'local'.
 
+ARG BASE_IMAGE=debian:bookworm-slim
```
for reproducibility reason, i suggest to keep the sha256 of the current latest image

AnXh3L0 replied:
Ok, adding back the sha256

evilaliv3 commented on docker/docker-compose.yml:
```diff
@@ -1,16 +1,147 @@
+      args:
+        INSTALL_FROM: local
+        DISTRIBUTION: bookworm
+        BASE_IMAGE: ${BASE_IMAGE_BOOKWORM:-debian:bookworm-slim}
```
@AnXh3L0: could we add the hash in any BASE_IMAGE defined in this file?

evilaliv3 commented on scripts/install.sh:
```diff
@@ -161,18 +218,56 @@ if [ -d /globaleaks/deb ]; then
+  fi
+  
+  # For remote installation, also prevent service restart in containers
+  if [ $DOCKER_ENV -eq 0 ] && ([ -f /.dockerenv ] || [ -n "${container}" ] || grep -q '/docker\|/lxc' /proc/1/cgroup 2>/dev/null); then
```
@AnXh3L0 : would you please explain to me this line and what it makes? when do we need lxc?

AnXh3L0 replied:
It checks if we are running inside a container, in case the first time we don't get a correct detection, sort of like a failsafe. The lxc is actually not necessary in this case, since I didn't know if any of our users ever deployed GlobaLeaks in lxc, so I assume we can probably remove it completely and only check for Docker instead.

evilaliv3 replied:
yes i think so, better to remove something that we are not controlling or that was explicitly requested by a user actively using it.

evilaliv3 reviewed (changes requested):
Thank you @AnXh3L0 

I think all is good to go. I would amiliorate just adding the hashes of all the base images used

AnXh3L0 committed:
Added sha256 pinned images and removed lxc line

AnXh3L0 committed:
Update hash on all lines

evilaliv3 commented:
Thank you @AnXh3L0 

Actually i notice still duplication in hash definition and hash usage.

As example i refer to:
```bash
BASE_IMAGE_BOOKWORM: debian:bookworm-slim@sha256:36e591f228bb9b99348f584e83f16e012c33ba5cad44ef5981a1d7c0a93eca22
```

and:
```bash
BASE_IMAGE: ${BASE_IMAGE_BOOKWORM:-debian:bookworm-slim@sha256:36e591f228bb9b99348f584e83f16e012c33ba5cad44ef5981a1d7c0a93eca22}
```

Above is defined an hash but while using it the hash is provided again. Is this necessary or could we write it once?

AnXh3L0 committed:
Removed fallback hashes

AnXh3L0 committed:
Changed envs

evilaliv3 committed:
Perform review of branch feature/gh-actions-fix

Changes applied:
- restored original static docker-compose.yml used for production purposes
- extended original docker-compose with health checks as in new testing compose
- add missing testing of debian bullseye
- reduced code duplication
- removed usage of deprecated apt-key

evilaliv3 committed:
Ensure to use the correct port based on the specified testing distribution

evilaliv3 reviewed (approved):
🚀 LGTM @AnXh3L0 !

In my review i've applied the following changes and attempted to reduce code duplication:
    - restored original static docker-compose.yml used for production purposes
    - extended original docker-compose with health checks as in new testing compose
    - add missing testing of debian bullseye
    - removed usage of deprecated apt-key

Proceeding with the merge!
